### ISSUE TITLE
[Linux] Distorted audio.

### ISSUE THREAD

**Issue by maxz** [luciusDXL/TheForceEngine#325]
The MIDI soundtrack sounds fine, but the voices, weapon sounds, ambient sounds and so on are heavily distorted on my Linux system using only ALSA.

Since the audio is played back, just in a very broken state and judging by the log, I guess it could be a bitrate mismatch due to RtAudio not finding the proper details of the underlying device.

```
[Startup] TFE_AudioSystem::init
[Audio] Attempting to initialize Audio System using API 'Linux ALSA'.
ALSA lib /var/tmp/portage/media-libs/alsa-lib-1.2.9/work/alsa-lib-1.2.9/src/confmisc.c:165:(snd_config_get_card) Cannot get card index for CODEC

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (default), No such device.


RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:2,0), Device or resource busy.

[Audio] Default Audio Output ID: 0.
[Audio] Device Count: 8.
ALSA lib /var/tmp/portage/media-libs/alsa-lib-1.2.9/work/alsa-lib-1.2.9/src/confmisc.c:165:(snd_config_get_card) Cannot get card index for CODEC

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (default), No such device.

[Audio] Device: 'default', ID: 0, Output Channel Count: 10000.
[Audio] Device: 'hw:Loopback,0', ID: 1, Output Channel Count: 32.
[Audio] Device: 'hw:Loopback,1', ID: 2, Output Channel Count: 32.
[Audio] Device: 'hw:HDA ATI HDMI,3', ID: 3, Output Channel Count: 2.

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:2,0), Device or resource busy.

[Audio] Device: 'hw:HD-Audio Generic,0', ID: 4, Output Channel Count: 0.
[Audio] Device: 'hw:HD-Audio Generic,1', ID: 5, Output Channel Count: 2.
[Audio] Device: 'hw:HD-Audio Generic,2', ID: 6, Output Channel Count: 0.
[Audio] Device: 'hw:HD Pro Webcam C920,0', ID: 7, Output Channel Count: 0.
[Audio] Audio initialized using API 'Linux ALSA'.
ALSA lib /var/tmp/portage/media-libs/alsa-lib-1.2.9/work/alsa-lib-1.2.9/src/confmisc.c:165:(snd_config_get_card) Cannot get card index for CODEC

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (default), No such device.


RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:2,0), Device or resource busy.

ALSA lib /var/tmp/portage/media-libs/alsa-lib-1.2.9/work/alsa-lib-1.2.9/src/confmisc.c:165:(snd_config_get_card) Cannot get card index for CODEC

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (default), No such device.

ALSA lib /var/tmp/portage/media-libs/alsa-lib-1.2.9/work/alsa-lib-1.2.9/src/confmisc.c:165:(snd_config_get_card) Cannot get card index for CODEC

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (default), No such device.

[Audio] Starting up audio stream for device 'default', id 0.
ALSA lib /var/tmp/portage/media-libs/alsa-lib-1.2.9/work/alsa-lib-1.2.9/src/confmisc.c:165:(snd_config_get_card) Cannot get card index for CODEC

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (default), No such device.


RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:2,0), Device or resource busy.

[Startup] TFE_MidiPlayer::init
[Startup] TFE_Polygon::init
```

**Comment by luciusDXL** [comment]
This was probably caused by a memory overwrite that was recently fixed. Can you try the latest in master?

**Comment by maxz** [comment]
No, I don't run PulseAudio or PipeWire.

**Comment by maxz** [comment]
It's still there in commit `d8e0809aa6ff1e765d900450e5a489eb367c14cf` (the current head of the master branch.)

```
[Startup] TFE_AudioSystem::init
[Audio] Attempting to initialize Audio System using API 'Linux ALSA'.
ALSA lib /var/tmp/portage/media-libs/alsa-lib-1.2.9/work/alsa-lib-1.2.9/src/confmisc.c:165:(snd_config_get_card) Cannot get card index for CODEC

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (default), No such device.


RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:2,0), Device or resource busy.

[Audio] Default Audio Output ID: 0.
[Audio] Device Count: 8.
ALSA lib /var/tmp/portage/media-libs/alsa-lib-1.2.9/work/alsa-lib-1.2.9/src/confmisc.c:165:(snd_config_get_card) Cannot get card index for CODEC

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (default), No such device.

[Audio] Device: 'default', ID: 0, Output Channel Count: 10000.
[Audio] Device: 'hw:Loopback,0', ID: 1, Output Channel Count: 32.
[Audio] Device: 'hw:Loopback,1', ID: 2, Output Channel Count: 32.
[Audio] Device: 'hw:HDA ATI HDMI,3', ID: 3, Output Channel Count: 2.

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:2,0), Device or resource busy.

[Audio] Device: 'hw:HD-Audio Generic,0', ID: 4, Output Channel Count: 0.
[Audio] Device: 'hw:HD-Audio Generic,1', ID: 5, Output Channel Count: 2.
[Audio] Device: 'hw:HD-Audio Generic,2', ID: 6, Output Channel Count: 0.
[Audio] Device: 'hw:HD Pro Webcam C920,0', ID: 7, Output Channel Count: 0.
[Audio] Audio initialized using API 'Linux ALSA'.
ALSA lib /var/tmp/portage/media-libs/alsa-lib-1.2.9/work/alsa-lib-1.2.9/src/confmisc.c:165:(snd_config_get_card) Cannot get card index for CODEC

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (default), No such device.


RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:2,0), Device or resource busy.

ALSA lib /var/tmp/portage/media-libs/alsa-lib-1.2.9/work/alsa-lib-1.2.9/src/confmisc.c:165:(snd_config_get_card) Cannot get card index for CODEC

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (default), No such device.

ALSA lib /var/tmp/portage/media-libs/alsa-lib-1.2.9/work/alsa-lib-1.2.9/src/confmisc.c:165:(snd_config_get_card) Cannot get card index for CODEC

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (default), No such device.

[Audio] Starting up audio stream for device 'default', id 0.
ALSA lib /var/tmp/portage/media-libs/alsa-lib-1.2.9/work/alsa-lib-1.2.9/src/confmisc.c:165:(snd_config_get_card) Cannot get card index for CODEC

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (default), No such device.


RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:2,0), Device or resource busy.

[Startup] TFE_MidiPlayer::init
```

**Comment by maxz** [comment]
I have a dmix PCM device set up. I have been using ALSA for decades on desktop Linux. And been burned by the early times of PulseAudio. I will never switch to PA because even today the CPU usage at times is absolutely ridiculous. But once I feel that PipeWire is ready, I will probably use PipeWire on top of ALSA.

My default device is an asym device (to combine the playback and microphone input because some software like browsers require it this way) which then routes into two dmix device, one for regular playback of what I hear and one for recording my sound.

And TFE does play audio. But it's very distorted e. g. slower and sounds like played back at a much lower bitrate.

Virtually all sound producing software works fine, even in parallel. Although there are some newer Unity games which don't support ALSA or the wrapper apulse and which I therefore have to play without sound.

**Comment by maxz** [comment]
Recorded playing Dark Forces in DOSBox: [dosbox.wav](https://github.com/maxz/tfe-audio-bug/blob/master/dosbox.wav)
Recorded in the latest version of TFE: [tfe.wav](https://github.com/maxz/tfe-audio-bug/blob/master/tfe.wav)

**Comment by luciusDXL** [comment]
Yep, it definitely should not sound like that. Is the framerate also really low? It sounds like the system is struggling to process the audio for some reason.

**Comment by maxz** [comment]
No, the framerate is constantly at 60 FPS. And the system way more powerful than required by the game. The processor load is really low.

I can also play all kinds of other audio in parallel which does not suffer.

Playing Dark Forces in TFE with Lapti Nek playing in parallel: [parallel-lapti-nek.wav](https://github.com/maxz/tfe-audio-bug/raw/master/parallel-lapti-nek.wav)

**Comment by maxz** [comment]
> Sounds like bad resampling. You wrote MIDI was fine: Is this the TFE built-in MIDI (SF2/OPL3) or System MIDI?

It's OPL3. I'm pretty sure system MIDI would also work, since I use that for other games, but I have not tested it.

The patch sadly does not change anything about the output.

**Comment by maxz** [comment]
The XZ compressed TFE binary: [theforceengine.xz](https://github.com/maxz/tfe-audio-bug/raw/master/theforceengine.xz)

```
# ~/.asoundrc

# Requires "alsa-plugins".
defaults.pcm.rate_converter samplerate_best

pcm.multi_playback {
  type route
  slave.pcm {
    type multi

    slaves.a.pcm dmixer_playback
    slaves.a.channels 2
    slaves.b.pcm dmixer_playback_loop
    slaves.b.channels 2

    bindings.0.slave a
    bindings.0.channel 0
    bindings.1.slave a
    bindings.1.channel 1

    bindings.2.slave b
    bindings.2.channel 0
    bindings.3.slave b
    bindings.3.channel 1
  }
  ttable.0.0 1.0
  ttable.1.1 1.0
  ttable.0.2 1.0
  ttable.1.3 1.0
}

# arecord --dump-hw-params -D hw:2,0
pcm.dmixer_playback {
  type dmix
  ipc_key 1
  ipc_perm 0600
  slave {
    pcm "hw:Generic,0"
    period_size 2048
    buffer_size 16384
    # Minimum rate: 44100 Hz; Maximum rate: 192000 Hz
    rate 48000
    channels 2
    format S32_LE
  }
  bindings {
    0 0
    1 1
  }
  slowptr true
}

pcm.dmixer_playback_loop {
  type dmix
  ipc_key 2
  ipc_perm 0600
  slave {
    pcm "hw:Loopback,0,0"
    period_size 2048
    buffer_size 16384
    # Minimum rate: 44100 Hz; Maximum rate: 192000 Hz
    rate 48000
    channels 2
    format S32_LE
  }
  bindings {
    0 0
    1 1
  }
  slowptr true
}

# Default control.
ctl.!default {
  type hw
  card "Generic"
}

# Default in- and output.
pcm.!default {
  # Important: Define playback AND capture.
  type asym
  playback.pcm {
    type plug
    slave.pcm {
      @func getenv
      vars [ ALSAPCM ]
      default multi_playback
    }
  }
  capture.pcm {
    type plug
    slave.pcm microphone
  }
}

pcm.microphone {
  type dsnoop
  ipc_key 3
  slave {
    pcm "hw:CODEC,0"
    period_size 2048
    buffer_size 16384
    # Minimum rate: 44100 Hz; Maximum rate: 192000 Hz
    rate 48000
    channels 2
    format S32_LE
  }
  bindings {
    0 0
    1 1
  }
  slowptr true
}

pcm.duplicated_output {
  type plug
  slave.pcm "hw:Loopback,1,0"
}
```

**Comment by maxz** [comment]
Without my configuration there is no sound at all because ALSA does not pick up the right device. Even when pointing it in the right direction with `defaults.pcm.card 2` and `defaults.ctl.card 2`.

If I disable the digital mixing by setting the default pcm directly to the device (which then obviously does not let me listen to more than one audio source at a time), the sound seems to work correctly.

```
[Startup] TFE_AudioSystem::init
[Audio] Attempting to initialize Audio System using API 'Linux ALSA'.
[Audio] Default Audio Output ID: 0.
[Audio] Device Count: 8.
[Audio] Device: 'default', ID: 0, Output Channel Count: 2.
[Audio] Device: 'hw:Loopback,0', ID: 1, Output Channel Count: 32.
[Audio] Device: 'hw:Loopback,1', ID: 2, Output Channel Count: 32.
[Audio] Device: 'hw:HDA ATI HDMI,3', ID: 3, Output Channel Count: 2.
[Audio] Device: 'hw:HD-Audio Generic,0', ID: 4, Output Channel Count: 2.
[Audio] Device: 'hw:HD-Audio Generic,1', ID: 5, Output Channel Count: 2.
[Audio] Device: 'hw:HD-Audio Generic,2', ID: 6, Output Channel Count: 0.
[Audio] Device: 'hw:HD Pro Webcam C920,0', ID: 7, Output Channel Count: 0.
[Audio] Audio initialized using API 'Linux ALSA'.
[Audio] Starting up audio stream for device 'default', id 0.
[Startup] TFE_MidiPlayer::init
```

Which still points in the direction that RtAudio does not pick up something correctly with my setup as written in my first comment.

**Comment by maxz** [comment]
Ok, so I tried to create a minimal asound.conf which works normally with other applications and shows this behaviour with TFE.

Setting the period size to 1024 leads to a segmentation fault in TFE on my system.
Setting it to 2 or 16384 leads to different sounding audio problems.

If you set no period_size, then TFE does not play back any audio, while mpv and Firefox work fine.

```
pcm.!default {
        type plug
        slave.pcm "dmixer"
}

pcm.dmixer {
           type dmix
           ipc_key 1
           slave {
                 pcm "hw:2,0"
                 period_size 1024                                                            
           }                                                              
}
```

```
[Progam Flow] The Force Engine Game Loop Started
[Game] Dark Forces Version: 1.0 (Build 1)
[Error : CrashHandler] Received Signal 11 errno 88 code 0
[Error : CrashHandler] faulting address 0x58
[Error : CrashHandler] Backtrace 8:
[Error : CrashHandler] 000 ./theforceengine(+0x1ad509) [0x57b0d9b80509]
[Error : CrashHandler] 001 /lib64/libc.so.6(+0x38e60) [0x7869b4964e60]
[Error : CrashHandler] 002 ./theforceengine(+0x6f2a2) [0x57b0d9a422a2]
[Error : CrashHandler] 003 ./theforceengine(+0x6e317) [0x57b0d9a41317]
[Error : CrashHandler] 004 /usr/lib64/librtaudio.so.6(_ZN9RtApiAlsa13callbackEventEv+0xbd) [0x7869b4eb1a9d]
[Error : CrashHandler] 005 /usr/lib64/librtaudio.so.6(+0x10c6d) [0x7869b4eb2c6d]
[Error : CrashHandler] 006 /lib64/libc.so.6(+0x8783a) [0x7869b49b383a]
[Error : CrashHandler] 007 /lib64/libc.so.6(+0x10a0ac) [0x7869b4a360ac]
Segmentation fault
```

**Comment by maxz** [comment]
I'm not sure whether there can be something done on the TFE side during the initialisation of RtAudio or whether it's purely up to RtAudio to fix this. I guess to write a proper bug report for RtAudio I'd have to produce a minimal test case which displays this behaviour with the library first.

Is there a reason for not using SDL2 for the audio part too?

**Comment by maxz** [comment]
> > ALSA lib /var/tmp/portage/media-libs/alsa-lib-1.2.9/work/alsa-lib-1.2.9/src/pcm/pcm_direct.c:2010:(snd1_pcm_direct_parse_open_conf) Invalid type for tstamp_type
> 
> That points to an error in one of your alsa config files, and at least here "tstamp_type" is only used in /usr/share/alsa/pcm/dmix.conf Do other SDL2-Audio using apps work? If so, which ones, so I can look at their audio code.

I'm not explicitly defining the tstamp_type in any of my ALSA configuration files.
I will try defining all of the possible values (default, gettimeofday, monotonic, monotonic_raw) and see whether that changes anything. I'd expect it to choose default if nothing is specified explicitly.

There are many commercial games that use SDL2 and work, but the source code for those is not public. (Turok, Dwarf Fortress on Steam and so on)
One which I used in the past and which has its source code available is [Barotrauma](https://github.com/Regalis11/Barotrauma), another one is [Xonotic](https://gitlab.com/xonotic/xonotic/). I just re-tested both to make sure that the sound works absolutely fine for me.
If you come across another one which uses SDL2 and which you'd like to test me I can do so.

Basically the only problems I ever had with my setup are with programs that only support PA, the current problems with TFE and rare problems with the latest version of Mumble, but only regarding some speakers and not everything. And problems with browser audio input until I had figured out that browsers expect an asym default PCM which combines output and input.

**Comment by maxz** [comment]
> I'm not explicitly defining the tstamp_type in any of my ALSA configuration files. I will try defining all of the possible values (default, gettimeofday, monotonic, monotonic_raw) and see whether that changes anything. I'd expect it to choose default if nothing is specified explicitly.

Setting an explicit tstamp_type does not change a thing. The same behaviour also appears with the minimal asound.conf that I posted.

**Comment by maxz** [comment]
> EDIT: I now had a look at the DarkPlaces engine sdl sound init code, and it's almost identical to mine, except for the requested format. As an experiment, in function "startOutput()" change "specin.format = AUDIO_F32LSB;" to AUDIO_S16 and see whether that gets the device initialized?

Sadly not:
```
[Startup] TFE_AudioSystem::init
[AudioSDL] SDLAudio using interface 'alsa'
[AudioSDL] Device 00: Loopback, Loopback PCM
[AudioSDL] Device 01: HD-Audio Generic, ALC887-VD Analog
[AudioSDL] Starting up audio stream for device 'HD-Audio Generic, ALC887-VD Analog'
ALSA lib /var/tmp/portage/media-libs/alsa-lib-1.2.9/work/alsa-lib-1.2.9/src/pcm/pcm_direct.c:2010:(snd1_pcm_direct_parse_open_conf) Invalid type for tstamp_type
[Error : AudioSDL] Open Audio Device HD-Audio Generic, ALC887-VD Analog failed with 'ALSA: Couldn't open audio device: Invalid argument'
[Error : Audio] Cannot start audio stream.
[Startup] TFE_MidiPlayer::init
```

If it really is functionally identical except for that part, then it would be really strange that I have no problem at all with xonotic-sdl and no sound with your SDL sound patch.

**Comment by maxz** [comment]
> Oh and change the 1st parameter of SDL_OpenAudioDevice() to NULL (instead of dn)

Doing so actually leads to at least some audio output.
But still with a lot of static mixed in.

```
[Startup] TFE_AudioSystem::init
[AudioSDL] SDLAudio using interface 'alsa'
[AudioSDL] Device 00: Loopback, Loopback PCM
[AudioSDL] Device 01: HD-Audio Generic, ALC887-VD Analog
[AudioSDL] Starting up audio stream for device 'HD-Audio Generic, ALC887-VD Analog'
[Startup] TFE_MidiPlayer::init
[Startup] TFE_Polygon::init
[Startup] TFE_Image::init
[Startup] TFE_FrontEndUI::init
[MemoryRegion] Allocated new memory block in region 'game' - new size is 1 blocks, total size is '8388608'
[MemoryRegion] Allocated new memory block in region 'level' - new size is 1 blocks, total size is '8388608'
[Progam Flow] The Force Engine Game Loop Started
[Game] Dark Forces Version: 1.0 (Build 1)
[MemoryRegion] Allocated new memory block in region 'Landru' - new size is 1 blocks, total size is '4194304'
[MemoryRegion] Allocated new memory block in region 'Cutscene' - new size is 1 blocks, total size is '8388608'
[Audio] Shutdown
[AudioSDL] Stop Audio Stream.
double free or corruption (!prev)
[Error : CrashHandler] Received Signal 6 errno 8806 code 0
```

BUT when building with NULL as first argument but with the original format instead of AUDIO_S16:
```
[Startup] TFE_AudioSystem::init
[AudioSDL] SDLAudio using interface 'alsa'
[AudioSDL] Device 00: Loopback, Loopback PCM
[AudioSDL] Device 01: HD-Audio Generic, ALC887-VD Analog
[AudioSDL] Starting up audio stream for device 'HD-Audio Generic, ALC887-VD Analog'
[Startup] TFE_MidiPlayer::init
```
And henceforth all audio worked correctly. Nice.

**Comment by maxz** [comment]
So we now got one solution working. Which is great (and SDL2 would be my preferred option anyways.)
For the sake of completeness I will also test RtAudio version 6.0.1 in the coming days and report back on my findings. I guess I should test whether the segmentation fault still exists in their latest version and at least report that one.

That was quite the elusive problem. And the number of people running a pure ALSA setup is probably very low.
If you got a more refined version of the SDL audio patch and want me to test it on my setup, just notify me.

**Comment by maxz** [comment]
Broken by default, but once I explicitly choose "<autoselect>" as audio device, it works.
And I guess the wrong device might only have been selected by default because I had explicitly selected it during one of my tries.

**Comment by maxz** [comment]
There was no empty string selection. It was merely a rendering issue that github had with `<autoselect>`.
But if the patch seems sensible to you then it would probably prevent other possible issues.

**Comment by luciusDXL** [comment]
The SDL audio changes have been pushed to master. Do these changes fix the distorted audio @maxz? If so, please close the issue.

**Comment by maxz** [comment]
I will test it again in the merged state and report back. If it does I'll close the issue.

**Comment by luciusDXL** [comment]
> I will test it again in the merged state and report back. If it does I'll close the issue.

Thanks. 👍

**Comment by maxz** [comment]
The SDL version works great.

I also wanted to test it with RtAudio version 6.0.1, but saw that RtAudio was fully removed from TFE. I checked out an older version of TFE to test it anyways and it turns out that they seem to have changed their API in version 6.0.1 and it does not even compile without various changes.

So this issue is fully resolved.
